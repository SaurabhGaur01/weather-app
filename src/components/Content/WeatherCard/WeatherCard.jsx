import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import ItemsCarousel from 'react-items-carousel';
import IconButton from '@material-ui/core/IconButton';
import ArrowForwardIosIcon from '@material-ui/icons/ArrowForwardIos';
import ArrowBackIosIcon from '@material-ui/icons/ArrowBackIos';
import { weatherCardClass } from '../../../utils/getStyles';
import WeatherChart from '../Chart/WeatherChart';
import { setChartData } from '../../../ducks/chartData';

const WeatherCard = ({ weatherData, unit, actionSetChartingData }) => {
  React.useEffect(() => {
    actionSetChartingData(weatherData[0].dataForChart);
  });
  const classes = weatherCardClass();
  const [activeItemIndex, setActiveItemIndex] = useState(0);
  const handleClick = (data) => {
    actionSetChartingData(data.dataForChart);
  }
  return (
    <React.Fragment>
      <div className={classes.card}>
        <div className={`weather-card ${classes.cardContainer}`}>
          <ItemsCarousel
            infiniteLoop={false}
            gutter={1}
            chevronWidth={60}
            numberOfCards={3}
            slidesToScroll={2}
            firstAndLastGutter={false}
            requestToChangeActive={setActiveItemIndex}
            activeItemIndex={activeItemIndex}
            outsideChevron ={true}
            rightChevron={
              <IconButton color="primary" aria-label="forward" component="span" >
                <ArrowForwardIosIcon color="primary"/>
              </IconButton>
            }
            leftChevron={
              <IconButton color="primary" aria-label="backward" component="span"
              >
                <ArrowBackIosIcon color="primary"/>
              </IconButton>
            }
          >  
            {
              weatherData.map( (day, index) => ( 
                <Card className={classes.root} key={index}>
                  <CardActionArea onClick={() => handleClick(day)}>
                    <CardContent>
                      <Typography variant="h6" component="h2">
                          {day.date}
                      </Typography>
                      <Typography className={classes.title} color="textSecondary">
                          Avg Temp:
                      </Typography>
                      <Typography variant="h6" component="h2">
                          {day.temperature.avgtemp} {unit}
                      </Typography>
                      <Typography className={classes.title} color="textSecondary">
                          Weather:
                      </Typography>
                      <Typography variant="subtitle2" gutterBottom>
                      {day.description}
                      </Typography>
                    </CardContent>
                  </CardActionArea> 
                </Card>
              ))
            }
          </ItemsCarousel>
        </div>
      </div>
      <WeatherChart />
    </React.Fragment>  
  )
}

WeatherCard.propTypes = {
  unit: PropTypes.string.isRequired,
  weatherData: PropTypes.arrayOf(PropTypes.shape({
    date: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    temperature: PropTypes.shape({
      avgtemp: PropTypes.string.isRequired,
    }),
    dataForChart: PropTypes.arrayOf(PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    })),
  })).isRequired,
  actionSetChartingData: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  weatherData: state.weatherData,
  unit: state.unit,
});

const mapDispatchAsProps = dispatch =>{
  return {
      actionSetChartingData: data => dispatch(setChartData(data)),
  };
}

const hocChain = compose(
  connect(mapStateToProps, mapDispatchAsProps),
);

export { WeatherCard as TestableWeatherCard };

export default hocChain(WeatherCard);