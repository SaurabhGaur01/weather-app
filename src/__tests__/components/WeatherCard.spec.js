import React from 'react';
import { shallow } from 'enzyme';
import { TestableWeatherCard } from '../../components/Content/WeatherCard/WeatherCard';

jest.mock('@material-ui/core/Card', () => 'Card');
jest.mock('@material-ui/core/CardActionArea', () => 'CardActionArea');
jest.mock('@material-ui/core/CardContent', () => 'CardContent');
jest.mock('@material-ui/core/Typography', () => 'Typography');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');
jest.mock('@material-ui/icons/ArrowForwardIos', () => 'ArrowForwardIos');
jest.mock('@material-ui/icons/ArrowBackIos', () => 'ArrowBackIos');

describe('WeatherCard Component', () => {
    let props;

    beforeEach(() =>{
        props = {
            actionSetChartingData: jest.fn(),
            handleClick: jest.fn(),
            unit: 'F',
            weatherData: [
                {
                    date: '',
                    description: '',
                    temperature: {
                        avgtemp: ''
                    },
                    dataForChart: [
                        {
                            text: '03:00',
                            value: 50.
                        },
                        {
                            text: '04:00',
                            value: 60.
                        },
                    ]
                }
            ]
        };
    });

    it('should render correctly', () => {
        const renderedModule = shallow(<TestableWeatherCard {...props} />);
        expect(renderedModule).toMatchSnapshot();
    });

    it('should call actionSetChartingData when component mounts', () => {
        jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());
        const { actionSetChartingData } = props;
        shallow(<TestableWeatherCard {...props} />);
        expect(actionSetChartingData).toHaveBeenCalledTimes(1);
    });
})