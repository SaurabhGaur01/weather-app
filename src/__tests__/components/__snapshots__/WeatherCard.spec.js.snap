// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCard Component should render correctly 1`] = `
<React.Fragment>
  <div
    className="makeStyles-card-3"
  >
    <div
      className="weather-card makeStyles-cardContainer-6"
    >
      <r
        activeItemIndex={0}
        activePosition="left"
        alwaysShowChevrons={false}
        calculateActualTranslateX={[Function]}
        chevronWidth={60}
        classes={Object {}}
        disableSwipe={false}
        enablePlaceholder={false}
        firstAndLastGutter={false}
        gutter={1}
        infiniteLoop={false}
        leftChevron={
          <IconButton
            aria-label="backward"
            color="primary"
            component="span"
          >
            <ArrowBackIos
              color="primary"
            />
          </IconButton>
        }
        numberOfCards={3}
        numberOfPlaceholderItems={0}
        onActiveStateChange={null}
        outsideChevron={true}
        placeholderItem={null}
        requestToChangeActive={[Function]}
        rightChevron={
          <IconButton
            aria-label="forward"
            color="primary"
            component="span"
          >
            <ArrowForwardIos
              color="primary"
            />
          </IconButton>
        }
        showSlither={false}
        slidesToScroll={2}
      >
        <Card
          className="makeStyles-root-1"
          key="0"
        >
          <CardActionArea
            onClick={[Function]}
          >
            <CardContent>
              <Typography
                component="h2"
                variant="h6"
              />
              <Typography
                className="makeStyles-title-2"
                color="textSecondary"
              >
                Avg Temp:
              </Typography>
              <Typography
                component="h2"
                variant="h6"
              >
                 
                F
              </Typography>
              <Typography
                className="makeStyles-title-2"
                color="textSecondary"
              >
                Weather:
              </Typography>
              <Typography
                gutterBottom={true}
                variant="subtitle2"
              />
            </CardContent>
          </CardActionArea>
        </Card>
      </r>
    </div>
  </div>
  <[object Object] />
</React.Fragment>
`;
